File: public/PublicMedia/static/solutions/elixir/problem5.ex
# Purpose: What is the smallest positive number that is evenly divisible by all of
# the numbers from 1 to 20?

# Author: Eben Quenneville
# Date: 2023-03-02
import ExUnit.Assertions

defmodule Problem5 do
  # Least common multiple of `a` and `b`
  def lcm(a, b) do
    div(a * b, Integer.gcd(a, b))
  end

  # Main driver
  def problem() do
    divisors = Enum.to_list(1..20)
    List.foldl(divisors, 1, fn x, acc -> lcm(acc, x) end)
  end

  # Check solution
  def main() do
    solution = problem()
    assert solution == 232_792_560
    IO.puts("Output: #{solution}")
  end
end

Problem5.main()

File: public/PublicMedia/static/solutions/clojure/problem5.clj
(comment 
  "Purpose: What is the smallest positive number that is evenly divisible by all of the numbers from 1 to 20?
   
   Author: Eben Quenneville
   Date: 2023-03-04")

; Greatest common divisor
(defn gcd "returns the greatest common divisor of n1 and n2" [n1 n2]
  (if (== n2 0)
    n1
    (gcd n2 (mod n1 n2))))

; Least common multiple
(defn lcm "returns the least common multiple of a and b" [a b]
  (/ (* a b) (gcd a b)))

;; Main driver code
(defn problem5 []
  ; find the lcm of 1 and 2, then 2 and 3, then 6 and 4, etc
  (reduce lcm (range 1 21)))

;; Assert the solution
(defn main []
  (assert (== (problem5) 232792560))
  (println "Output: " (problem5)))
(main)
